src/R/Components/Impl/Plots/Implementation/RPlotManager.cs

```diff
@@ -188,35 +283,143 @@ internal class RPlotManager : IRPlotManager {
+        private async Task ProcessPlotMessage(IRPlotDeviceViewModel viewModel, PlotMessage plot) {
+            if (plot.IsClearAll) {
+                await viewModel.PlotMessageClearAllAsync(plot.DeviceId, plot.DeviceNum);
+                History.RemoveAll(plot.DeviceId);
+            } else if (plot.IsPlot) {
+                try {
+                    var img = plot.ToBitmapImage();
+                    await viewModel.PlotMessageLoadPlotAsync(plot.DeviceId, plot.PlotId, img, plot.DeviceNum, plot.ActivePlotIndex, plot.PlotCount);
+                    History.AddOrUpdate(viewModel.DeviceName, viewModel.DeviceId, plot.PlotId, viewModel.SessionProcessId, img);
+                } catch (Exception e) when (!e.IsCriticalException()) {
+                    await viewModel.PlotMessageLoadErrorAsync(plot.DeviceId, plot.PlotId, plot.DeviceNum, plot.ActivePlotIndex, plot.PlotCount);
+                }
+            } else if (plot.IsError) {
+                await viewModel.PlotMessageLoadErrorAsync(plot.DeviceId, plot.PlotId, plot.DeviceNum, plot.ActivePlotIndex, plot.PlotCount);
+            }
+        }
 
-        public Task ExportToPdfAsync(string outputFilePath) =>
-            ExportAsync(outputFilePath, _interactiveWorkflow.RSession.ExportToPdfAsync(outputFilePath, PixelsToInches(_lastPixelWidth), PixelsToInches(_lastPixelHeight)));
+        private async Task RemoveFromHistoryAsync(Guid plotId) {
+            await InteractiveWorkflow.Shell.SwitchToMainThreadAsync();
+            History.Remove(plotId);
+        }
 
-        public void EndLocatorMode() {
-            EndLocatorMode(LocatorResult.CreateNotClicked());
+        private async Task ClearHistoryAsync() {
+            await InteractiveWorkflow.Shell.SwitchToMainThreadAsync();
+            History.Clear();
         }
 
-        public void EndLocatorMode(LocatorResult result) {
-            var tcs = _locatorTcs;
-            _locatorTcs = null;
-            tcs?.SetResult(result);
-            _interactiveWorkflow.Shell.DispatchOnUIThread(() => SetLocatorModeUI(false));
-            LocatorModeChanged?.Invoke(this, null);
+        private async void RSession_Mutated(object sender, EventArgs e) {
+            await RSessionMutatedAsync();
         }
 
-        private void SetLocatorModeUI(bool locatorMode) {
-            if (VisualComponent != null) {
-                VisualComponent.SetLocatorMode(locatorMode);
+        private async Task RSessionMutatedAsync() {
+            try {
+                var deviceId = await InteractiveWorkflow.RSession.GetActivePlotDeviceAsync();
+                var deviceChanged = ActiveDeviceId != deviceId;
+                ActiveDeviceId = deviceId;
+
+                // Update all the devices in parallel
+                var visualComponents = _visualComponents.Values.ToArray();
+                var tasks = visualComponents.Select(v => v?.ViewModel.RefreshDeviceNameAsync());
```

Comment:
To clarify, if you don't await on the call (and ultimately on the underlying `SendNotificationAsync` call in the host), all these packets are going to be pushed down the websocket connection as soon as they come in, without waiting for ACK from the other side between sends. So there isn't really any benefit derived from making special-purpose batch messages, since you can always batch any two or more messages with `Task.WhenAll` like this (and better yet, you can batch different kinds of requests that way).